File: snippets-python/shared.py
from typing import Any

from multiversx_sdk import (
    Address,
    ApiNetworkProvider,
    SmartContractController,
    SmartContractQueryResponse,
    SmartContractTransactionsFactory,
    Transaction,
    TransactionComputer,
    TransactionsFactoryConfig,
    UserSigner
)

from constants import API_URL, CHAIN_ID, EXPLORER_URL


def recall_nonce(transaction: Transaction):
    network_provider = create_network_provider()
    account = network_provider.get_account(transaction.sender)
    transaction.nonce = account.nonce


def sign_transaction(transaction: Transaction, signer: UserSigner):
    computer = TransactionComputer()
    bytes_for_signing = computer.compute_bytes_for_signing(transaction)
    transaction.signature = signer.sign(bytes_for_signing)


def broadcast_transaction(transaction: Transaction):
    network_provider = create_network_provider()
    hash = network_provider.send_transaction(transaction)

    print("See transaction:")
    print(f"{EXPLORER_URL}/transactions/{hash}")


def query_contract(address: Address, function: str, arguments: list[Any]) -> SmartContractQueryResponse:
    sc_controller = SmartContractController(chain_id=CHAIN_ID, network_provider=create_network_provider())
    query = sc_controller.create_query(
        contract=address,
        function=function,
        arguments=arguments
    )

    response = sc_controller.run_query(query)
    return response


def create_smart_contract_transactions_factory():
    factory_config = TransactionsFactoryConfig(CHAIN_ID)
    factory = SmartContractTransactionsFactory(factory_config)
    return factory


def create_network_provider():
    return ApiNetworkProvider(API_URL)

File: snippets-javascript/shared.js
const fs = require("fs");
const { ApiNetworkProvider } = require("@multiversx/sdk-network-providers");
const { SmartContractTransactionsFactory } = require("@multiversx/sdk-core/out/transactionsFactories");
const { TransactionsFactoryConfig } = require("@multiversx/sdk-core/out/transactionsFactories");
const { TokenComputer } = require("@multiversx/sdk-core");
const { AbiRegistry } = require("@multiversx/sdk-core");
const { API_URL, CHAIN_ID, EXPLORER_URL } = require("./constants");

function applyChainID(transaction) {
    transaction.setChainID(CHAIN_ID);
}

async function recallNonce(transaction) {
    const networkProvider = createNetworkProvider();
    const sender = transaction.sender;
    const account = await networkProvider.getAccount(sender);
    transaction.nonce = account.nonce;
}

async function signTransaction(transaction, signer) {
    const bytesForSigning = transaction.serializeForSigning();
    transaction.signature = await signer.sign(bytesForSigning);
}

async function broadcastTransaction(transaction) {
    const networkProvider = createNetworkProvider();
    const hash = await networkProvider.sendTransaction(transaction);

    console.log("See transaction:");
    console.log(`${EXPLORER_URL}/transactions/${hash}`);
}

function queryContract(address, func, arguments) {
}

function createSmartContractTransactionsFactory(abiFilePath) {
    const abiFileContent = fs.readFileSync(abiFilePath, { encoding: "utf8" });
    const abiObject = JSON.parse(abiFileContent);
    const abiRegistry = AbiRegistry.create(abiObject);

    const factoryConfig = new TransactionsFactoryConfig(CHAIN_ID);
    const factory = new SmartContractTransactionsFactory({
        config: factoryConfig,
        abi: abiRegistry,
        tokenComputer: new TokenComputer()
    });

    return factory;
}

function createNetworkProvider() {
    return new ApiNetworkProvider(API_URL);
}

module.exports = {
    applyChainID,
    recallNonce,
    signTransaction,
    broadcastTransaction,
    queryContract,
    createSmartContractTransactionsFactory
};
